x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -197,15 +213,29 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
             searchResponse = null;
         } else {
             /*
+             * MP TODO: UPDATE THIS DOCUMENTATION TO MATCH NEW LOGIC
              * Build the response, reducing aggs if we haven't already and
              * storing the result of the reduction, so we won't have to reduce
              * the same aggregation results a second time if nothing has changed.
              * This does cost memory because we have a reference to the finally
              * reduced aggs sitting around which can't be GCed until we get an update.
              */
-            InternalAggregations reducedAggs = reducedAggsSource.get();
-            reducedAggsSource = () -> reducedAggs;
-            searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            if (searchResponseMerger == null) {
+                System.err.println("XXX DEBUG 1 - no search response merger - local aggs only");
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            } else if (finalReduce == false) {
+                System.err.println("XXX DEBUG 2 using SRM + partial local results");
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                // TODO: check that this just returns the partial aggs response passed in if SearchResponseMerger
+                // has no other full SearchResponses
+                searchResponse = searchResponseMerger.getMergedResponse(clusters, buildResponse(task.getStartTimeNanos(), reducedAggs));
```

Comment:
You could hold a reference to ` Supplier<SearchResponse>` instead to that we do the same as we do for aggs: delay reduction until results are requested, hold the reduced results in memory for later retrieval once reduced for the first time.

Reply:
I don't think I follow.

First, this does hold the `reducedAggsSource` as an instance variable just like the current approach in main.

Second, the SearchResponse is not currently cached (in main) - it is always regenerated on demand and assigned only to a local variable. That is the same in my new impl.

Third, I don't think we want to cache a SearchResponse - that's a key reason for the reorg of the code in my last commit - to not hold more RAM in an instance var.

Fourth, I don't think we want to cache SearchResponses because the call to SearchResponseMerger may generate a different response if more clusters have reported in. We need to always generate a new response until the final response is written to disk/index.

I have a feeling that none of my above comments are addressing what you suggest though, but I'm unclear on what value a `Supplier<SearchResponse>` brings here. We already effectively have that in a SearchResponseMerger reference. Why does another layer of indirection help?